### PR TITLE
 Add support for allowing comments within Json while reading

### PR DESCRIPTION
**Pull request by ahsonkhan**
**TODO:**
- Should we just ignore comments or should we support users actually reading the comment values?

### CODE REVIEW

File: src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader_MultiSegment.cs [hunk 1]
```diff
@@ -377,14 +409,133 @@ private bool ConsumeValue(ref BufferReader<byte> reader, byte marker)
...
+            _lineNumber++;
+        Done:
+            _stack.Push((InternalJsonTokenType)TokenType);
+            TokenType = JsonTokenType.Comment;
+            CurrentIndex += 2 + Value.Length;
+            return true;
+        }
+
+        // TODO: Fix allocations
+        private bool ConsumeMultiLineComment(ref BufferReader<byte> reader)
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader_MultiSegment.cs]: Oh actually, never mind, I can use this overload:
`TryReadTo(out ReadOnlySequence<T> sequence, ReadOnlySpan<T> delimiter, bool advancePastDelimiter = true)`

I was looking for the API that returned a `ReadOnlySpan<T>` as an out. We seem to be mssing that one.

**Comment by ahsonkhan** [comment]
cc @KrzysztofCwalina, fyi

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs [hunk 2]
```diff
@@ -441,6 +442,9 @@ public unsafe bool TryReadTo(out ReadOnlySequence<T> sequence, ReadOnlySpan<T> d
                 ReadOnlySpan<T> next = Peek(peekBuffer);
                 if (next.SequenceEqual(delimiter))
                 {
+                    //TODO: Figure out a faster way to do this, potentially by avoiding the Advance in the previous TryReadTo call
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs]: Fyi @JeremyKuhne, noticed a bug. I don't think this is the cleanest solution and there are likely other edge cases.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs [hunk 3]
```diff
@@ -149,6 +149,7 @@ public void Rewind(long count)
             if (CurrentSpanIndex >= count)
             {
                 CurrentSpanIndex -= (int)count;
+                _moreData = true;
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs]: Thanks for writing the test for it

File: src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader_MultiSegment.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader_MultiSegment.cs]: Yeah, oversight. I think I've got it in the proposal.

**Pull request merged**